### PR TITLE
DPL: allow custom orbit when using timer / enumeration

### PR DESCRIPTION
**Pull request by ktf**
The firstTForbit field is now given by the formula:

firstTForbit = orbit-offset-enumeration + orbit-multiplier-enumeration + dh.tfCounter

for the enumeration and timer cases. One can use, e.g.:

--orbit-offset-enumeration and --orbit-multiplier-enumeration

on the command line to affect the two parameters.

### CODE REVIEW

**Comment by ktf** [comment]
@shahor02 I still need to fix the firstTFOrbit for the dummy, but this should give you enough flexibility for the time / enumeration case.

**Comment by shahor02** [comment]
@ktf Thanks! Actually, I need to set it not from the CL but in WorkflowSpec definition according to parameters extracted from the GRP. Since the MC workflows always start by digit reader, I guess the following should work ?
````
uint32_t firstOrbit, nOrbitsPerTF;
/// extract from values
readerSpec.options.push_back(ConfigParamSpec{"orbit-offset-enumeration", VariantType::Int64,  firstOrbit, {"initial value for the orbit"}});
readerSpec.options.push_back(ConfigParamSpec{"orbit-multiplier-enumeration", VariantType::Int64, nOrbitsPerTF, {"multiplier to get the orbit from the counter"}}
````
Will test tomorrow.

**Comment by ktf** [comment]
Ok. Yes it should work if you do that. It looks like I forgot one warning. Will check later.

File: Framework/Core/src/WorkflowHelpers.cxx [hunk 1]
```diff
@@ -289,6 +291,8 @@ void WorkflowHelpers::injectServiceDevices(WorkflowSpec& workflow, ConfigContext
     std::string prefix = "internal-dpl-";
     if (processor.inputs.empty() && processor.name.compare(0, prefix.size(), prefix) != 0) {
       processor.inputs.push_back(InputSpec{"enumeration", "DPL", "ENUM", static_cast<DataAllocator::SubSpecificationType>(compile_time_hash(processor.name.c_str())), Lifetime::Enumeration});
+      processor.options.push_back(ConfigParamSpec{"orbit-offset-enumeration", VariantType::Int64, 0ll, {"initial value for the enumeration"}});
+      processor.options.push_back(ConfigParamSpec{"orbit-multiplier-enumeration", VariantType::Int64, 1ll, {"initial value for the enumeration"}});
```
Comment (shahor02) [inline Framework/Core/src/WorkflowHelpers.cxx]: @ktf I've tried with lines below in the `its-reco-workflow.cxx`
````
auto wf = o2::its::reco_workflow::getWorkflow(...);
for (auto& spec : wf) {
   if (spec.inputs.empty()) {
      LOG(INFO) << "Configuring 1st orbit of " << spec.name;
      spec.options.push_back(ConfigParamSpec{"orbit-offset-enumeration", VariantType::Int64, orbitFirst , {"1st injected orbit"}});
      spec.options.push_back(ConfigParamSpec{"orbit-multiplier-enumeration", VariantType::Int64, nHBFPerTF), {"orbits/TF"}});
   }
}
````
but it leads to 
````
[INFO] Configuring 1st orbit of its-digit-reader
...
[ERROR] error while setting up workflow: 
Throw location unknown (consider using BOOST_THROW_EXCEPTION)
Dynamic exception type: boost::wrapexcept<boost::program_options::ambiguous_option>
std::exception::what: option is ambiguous and matches different versions of 'orbit-multiplier-enumeration'
````

Indeed, in the `o2-its-reco-workflow -h full ` I see duplicated options
````
  --orbit-offset-enumeration arg (=60)         1st injected orbit
  --orbit-multiplier-enumeration arg (=128)  orbits/TF
  --orbit-offset-enumeration arg (=0)            initial value for the enumeration
  --orbit-multiplier-enumeration arg (=1)       initial value for the enumeration
````

Could you add these 2 options after making sure that device has no them already ?
Also, the options hints should be corrected.

Comment (ktf) [reply Framework/Core/src/WorkflowHelpers.cxx]: yes, of course. Looking into it.

Comment (ktf) [reply Framework/Core/src/WorkflowHelpers.cxx]: Should be fixed now. Happy Easter! :-)

File: Framework/Core/src/WorkflowHelpers.cxx [hunk 2]
```diff
@@ -263,6 +264,8 @@ void WorkflowHelpers::injectServiceDevices(WorkflowSpec& workflow, ConfigContext
     {ConfigParamSpec{"aod-file", VariantType::String, {"Input AOD file"}},
      ConfigParamSpec{"aod-reader-json", VariantType::String, {"json configuration file"}},
      ConfigParamSpec{"time-limit", VariantType::Int64, 0ll, {"Maximum run time limit in seconds"}},
+     ConfigParamSpec{"orbit-offset-enumeration", VariantType::Int64, 0ll, {"initial value for the orbit"}},
+     ConfigParamSpec{"orbit-multiplier-enumeration", VariantType::Int64, 1ll, {"multiplier to get the orbit from the counter"}},
```
Comment (shahor02) [inline Framework/Core/src/WorkflowHelpers.cxx]: should not default be 0 also here, i.e. 1st orbit will stay 0?

**Comment by shahor02** [comment]
Thanks and Happy Easter!

Now it almost works, but I see 2 different problem 

1) Not related to this PR but to the firstTForbit in general: running a workflow starting with `internal-dpl-clock`, e.g. like
image
I see that the `DataHeader::firstTForbit` and `tfCounter` are not propagated correctly from the `DPL/ENUM/XXX` message.
E.g.  by hacking the `its-digit-reader`, `its-clusterer` and `its-cooked-tracker` to show 
````
  auto dh = DataRefUtils::getHeader<o2::header::DataHeader*>(pc.inputs().getByPos(0));
  LOG(INFO) << "1stOrbit : " << dh->firstTForbit << " TFid: " << dh->tfCounter << " for " << dh->dataOrigin.str << "/" << dh->dataDescription.str;
````
I get the following:
````
o2-its-reco-workflow --disable-mc --run | tee xx.log
 grep '1stOrbit' xx.log
[1793329:its-digit-reader]: [01:18:21][INFO] 1stOrbit : 60 TFid: 0 for DPL/ENUM
[1793330:its-clusterer]: [01:18:21][INFO] 1stOrbit : 4294967295 TFid: 4294967295 for ITS/DIGITS
[1793332:its-cooked-tracker]: [01:18:22][INFO] 1stOrbit : 4294967295 TFid: 4294967295 for ITS/CLUSTERSROF
````

2) certainly related to this PR: if I run a workflow like
````
o2-raw-file-reader-workflow --detect-tf0 --input-conf raw/ITS/ITSraw.cfg | o2-itsmft-stf-decoder-workflow
````
image

I get an exception:
````
[1855511:its-stf-decoder]: [02:32:10][STATE] Starting FairMQ state machine --> IDLE
[1855511:its-stf-decoder]: [02:32:10][STATE] IDLE ---> INITIALIZING DEVICE
[1855511:its-stf-decoder]: [02:32:10][STATE] Exiting FairMQ state machine
[1855511:its-stf-decoder]: [02:32:10][ERROR] Unhandled std::exception reached the top of main: bad type for option: orbit-offset, device shutting down.
````

I don't understand why the `orbit-offset` option pops up in the `its-stf-decoder` device since it has its proper inputs.
The workflow somehow gets options screwed up: if I run only 1st part, it shows that the device has an options and 
````
o2-raw-file-reader-workflow --detect-tf0 --input-conf raw/ITS/ITSraw.cfg -h
...
Available data processors (full info with '--help full'):
  --internal-dpl-clock arg              No processor options
  --raw-file-reader arg                 Use '--help raw-file-reader' to display processor options
````
and the same with  ` -h raw-file-reader` shows as expected:
````
Data processor options: raw-file-reader:
  --orbit-offset-enumeration arg (=0)   1st injected orbit
  --orbit-multiplier-enumeration arg (=1) orbits/TForbit
  --start-value-enumeration arg (=0)    initial value for the enumeration
  --end-value-enumeration arg (=-1)     final value for the enumeration
  --step-value-enumeration arg (=1)     step between one value and the other
````

But the full piped workflow help shows:
````
o2-raw-file-reader-workflow --detect-tf0 --input-conf raw/ITS/ITSraw.cfg | o2-itsmft-stf-decoder-workflow  -h
...
Available data processors (full info with '--help full'):
  --internal-dpl-clock arg              No processor options
  --raw-file-reader arg                 No processor options
  --its-stf-decoder arg                 Use '--help its-stf-decoder' to display processor options
````
i.e. the options of my 1st device, `raw-file-reader` (which has not input thus gets your new options added), have disappeared.
And the help of `its-stf-decoder` (which produces this exception), does not show the option like `orbit-offset`, due to which the exception is produced...

I have the data needed for these workflows here: https://cernbox.cern.ch/index.php/s/DWoZdpTubY5yBDP

**Comment by ktf** [comment]
I have corrected the name of the parameter. For the propagation I will need to do a bit more tests.

**Comment by shahor02** [comment]
Both problems are still there.

**Comment by shahor02** [comment]
@ktf: concerning the 1st problem (not relaying the DataHeader content from the `internal-dpl-clock` to downstream devices: in the gdb I see that all devices immediately following the `internal-dpl-clock` see always `info.state == InputChannelState::Pull`, so the line https://github.com/AliceO2Group/AliceO2/blob/1e26edb13791a8268817c4341a46addebd1dfd34/Framework/Core/src/DataProcessingDevice.cxx#L514-L515 prevents from executing https://github.com/AliceO2Group/AliceO2/blob/1e26edb13791a8268817c4341a46addebd1dfd34/Framework/Core/src/DataProcessingDevice.cxx#L545 which eventually invokes DataRelayer. Next level devices do see `info.state == InputChannelState::Running` so they relay the DH content w/o problems.

**Comment by shahor02** [comment]
... and for the problem (2): I see that the exception is thrown from here https://github.com/ktf/AliceO2/blob/485dbf67589d95d79d1812fe2fcbfb95ba8a13bd/Framework/Core/src/DeviceSpecHelpers.cxx#L240-L242 but, as I wrote above, this device is not supposed to have these options, they are added only to input-less devices. Are you sure these lines are correct? In any case, I guess the "expired message" stuff is activated because the preceding device (`raw-file-reader`) is loses all its options and the data is not processed correctly. This should be the primary reason.

**Comment by ktf** [comment]
The `internal-dpl-clock` is not a real device and it does not send anything. It's there just as a placeholder for the topology to be without dangling outputs, so it's normal that there is no "handleData" invoked. I am checking...

File: Framework/Core/src/WorkflowHelpers.cxx (hunk 2 above)
Comment (ktf) [reply Framework/Core/src/WorkflowHelpers.cxx]: yes, I changed it to 0 as discussed to indicate that the value was not provided.

**Comment by ktf** [comment]
@shahor02 I think I've fixed the issues you had with the exception as well. Could you verify?

The only remaining issue seems to be the QC functional test, although I do not understand why, yet.

**Comment by ktf** [comment]
It should also fix the orbit for Lifetime::Optional

**Comment by shahor02** [comment]
@ktf thanks, now both problems are gone. Concerning the disappearing options of the `raw-reader` (still there), it might be problem of printing or they are removed after the options are processed and the workflow is built. At least, the device does accept and properly accounts the options.

**Comment by ktf** [comment]
I think I tracked down the issue with options (at least for QC) to the fact timers have no relationship with orbit (since startTime is really a counter nowadays). So orbit will be set to 0 for the inputs associated to a timer.

**Review by shahor02 — APPROVED** [review]
For the raw-reader options loss: I've dared to check in the dev and see it also there. So, it is not related this PR, I've erroneously associated them due to the exception happening in the next to the reader device and mentioning nonexistent options.
As the exception is fixed, for me it is ok to merge this.

**Comment by ktf** [comment]
@davidrohr I see:

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/5852/0/Detectors/ITSMFT/common/reconstruction/src/RawPixelDecoder.cxx:21:10: error: 'omp.h' file not found [clang-diagnostic-error]
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/5852/0/Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:18:10: error: 'omp.h' file not found [clang-diagnostic-error]
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/5852/0/Detectors/Vertexing/src/SVertexer.cxx:22:10: error: 'omp.h' file not found [clang-diagnostic-error]
```

which is probably unrelated to this PR. Do you know what changed?

@TimoWilken could you fix the regular expressions so that `: error:` is shown in the summary?

**Comment by davidrohr** [comment]
@ktf : For the omp problems. This is an artifact of how the codechecker works. But in any case, it is not a real problem, since there is a filter afterwards. The codechecker only fails, if there are errors listed at the end below `========== List of errors found ==========`, which is empty for you.

In particular, @TimoWilken : We must not grep for `: error:` since that will yield false positives!!!

The final failure of this PR is due to `Error in <TFile::TFile>: file o2sim_Kine.root does not exist`, which is this error that has been popping up and disappearing again since few weeks in the full system test. I have now idea what causes it, and I cannot reproduce it locally running in the docker container locally. I have discussed with @TimoWilken whether the disk may run full, but we have no real logs on that, but he assumes it is not the case.
After discussion with @sawenzel , I have added additional debug output for the failing simulation part in alisw/alidist#2952 , let's see if that will help understand this.

**Comment by ktf** [comment]
Ok, I will merge this then. I am surprised there is no way to tell clang-tidy to ignore the WITH_OMP part of the `#ifdef`.

**Pull request merged**